### PR TITLE
Provides auto-init feature and documents operating modes

### PR DESCRIPTION
**Pull request by lorengordon**
Testing user mode:

1. Navigate to a directory where `tardigrade-ci/` is not in the parent/sibling path.
2. Create a test directory and change to it: `mkdir test-project && cd test-project`
3. Create a Makefile with the contents from tardigrade-ci README, but with a URL that points to my fork:
    ```
    echo 'include $(shell test -f .tardigrade-ci || curl -sSL -o .tardigrade-ci "https://raw.githubusercontent.com/lorengordon/tardigrade-ci/init-updates/bootstrap/Makefile.bootstrap"; echo .tardigrade-ci)' > Makefile`
    ```
4. Run `make help`, setting the envs to point at my fork/branch:
    ```
    $ TARDIGRADE_CI_ORG=lorengordon TARDIGRADE_CI_BRANCH=init-updates make help
    Cloning https://github.com/lorengordon/tardigrade-ci.git?ref=init-updates...
    Cloning into 'tardigrade-ci'...
    remote: Enumerating objects: 63, done.
    remote: Counting objects: 100% (63/63), done.
    remote: Compressing objects: 100% (50/50), done.
    remote: Total 63 (delta 24), reused 26 (delta 10), pack-reused 0
    Receiving objects: 100% (63/63), 54.75 KiB | 2.74 MiB/s, done.
    Resolving deltas: 100% (24/24), done.
    Available targets:

      cfn/lint                            Lints CloudFormation files
      ...
    ```

Testing developer mode:

1. Clone my fork (or add my remote to an existing clone), and checkout the `init-updates` branch
2. Create the `test-project` directory as a sibling directory to the tardigrade-ci clone
3. Follow the user mode instructions above for step 3 and 4. No need to set the envs in step 4 because the git repo is already setup for it. Note the init/clone is skipped.
    ```
    $ make help
    Available targets:

      cfn/lint                            Lints CloudFormation files
      ...
    ```

### CODE REVIEW

File: bootstrap/Makefile.bootstrap [hunk 1]
```diff
@@ -2,19 +2,64 @@ export SHELL = /bin/bash
...
+elif [[ \
+  "$(TARDIGRADE_CI_PATH)" == "$(TARDIGRADE_CI_PATH_LOCAL)" && \
+  -f "$(TARDIGRADE_CI_MAKEFILE)" && \
+  $$(git -C $(TARDIGRADE_CI_PATH_LOCAL) ls-remote "$(TARGIDRADE_CI_CLONE_URL)" "$(TARDIGRADE_CI_BRANCH)" | cut -f1) == $$(git -C $(TARDIGRADE_CI_PATH_LOCAL) rev-parse HEAD) \
+]]; then \
+  : ;\
+else \
+  $($(TARDIGRADE_CI_PROJECT)/clone) ;\
+fi
+endef
```
Comment (kbalk) [inline bootstrap/Makefile.bootstrap:41]: endef -> endif; in more than one place.

Comment (lorengordon) [reply bootstrap/Makefile.bootstrap:46]: I don't know why they chose this marker, it is not intuitive, and easily looks like a typo, but `endef` is the closing syntax for the Make `define` directive. 

* https://www.gnu.org/software/make/manual/html_node/Multi_002dLine.html

File: README.md [hunk 2]
```diff
@@ -45,58 +45,140 @@ This project can be utilized one of two ways, via docker or via a Makefile inclu
...
+specifying a version). The user may override the default behavior by providing
+the env explicitly, e.g. `TARDIGRADE_CI_BRANCH=foo make help`.
+
+In **developer mode**, the user is responsible for cloning the `tardigrade-ci`
+repo to a sibling directory of the current project, e.g. `/{project}/` and
+`/tardigrade-ci/`. When the bootstrap Makefile detects this condition, it will
+include the tardigrade-ci Makefile from the user-managed clone. This makes it easier
+for a developer to edit the tardigrade-ci project, and test those changes from
+a calling project.
+
```
Comment (eemperor) [inline README.md:174]: > When the bootstrap Makefile detects this condition...

I'm unclear a bit as to how the detection happens in **developer mode**.  With `tardigrade-ci` being a sibling directory, would the simple Makefile of the `/{project}/` be different and reference the local `tardigrade-ci` clone?  Or do you use the same simple Makefile and the bootstrap logic handles it.  I may need to better understand the bootstrap logic changes...

**Comment by kbalk** [comment]
Is CHANGELOG.md still needed?

File: README.md (hunk 2 above)
Comment (lorengordon) [reply README.md:174]: The detection logic/method is the same whether its developer or user mode. The bootstrap Makefile handles it, so the {project} Makefile does not change between developer/user mode. The flow is like this:

1. The {project} Makefile "includes" the bootstrap Makefile.
2. The bootstrap Makefile "includes" the tardigrade-ci Makefile, and has the logic to determine where the tardigrade-ci Makefile is. That location/path is what differentiates the user vs developer modes.

This flow is what I attempted to describe in the first two paragraphs of the new section "Makefile operating modes". Further suggestions welcome on how to clarify that language!

Comment (lorengordon) [reply README.md:174]: This developer mode vs user mode is not really a new thing. It just wasn't documented before. So with or without the changes in this PR, these modes are the same.

What changes in this PR is that in **user mode**, previously, you had to run `make init` manually before running any of the tardigrade-ci Makefile targets. With this PR, that is automatic. The bootstrap determines whether it needs to run the init logic, and does so if needed, as part of the "include" directive.

There are no changes to how **developer mode** works in this PR, I don't think.

**Comment by lorengordon** [comment]
Yes, CHANGELOG.md is still needed. We need to be better about updating it though.

File: README.md (hunk 2 above)
Comment (eemperor) [reply README.md:174]: Got it.  That explanation helped.

File: README.md [hunk 3]
```diff
@@ -45,58 +45,140 @@ This project can be utilized one of two ways, via docker or via a Makefile inclu
...
+Makefile and processes its directives and targets.
+
+The bootstrap Makefile is very barebones. Its primary purpose is to bootstrap this
+`tardigrade-ci` project so all the make targets are available. It does this through
+its own `include` directive. The bootstrap include directive processes a shell
+command with logic to determine whether to clone this project, and then echoes
+back the path to this project's Makefile.
+
+The bootstrap Makefile supports two "operating modes" that we're calling **user mode**
+and **developer mode**.
```
Comment (eemperor) [inline README.md:155]: Maybe reorder the explanation like below?

```suggestion
its own `include` directive.

The bootstrap Makefile supports two "operating modes" that we're calling **user mode** and **developer mode** 

The bootstrap include directive processes a shell command with logic to determine whether to operate in **user mode** or **developer mode**, and then echoes back the path to this project's Makefile.
```

File: README.md [hunk 4]
```diff
@@ -45,58 +45,140 @@ This project can be utilized one of two ways, via docker or via a Makefile inclu
...
+or simply do not invoke targets that rely on these packages:
+
+* `zip`
+* `nodejs`
+
+## Makefile "operating modes"
+
+When using the Makefile approach, the `include` directive is a shell command that
+retrieves the "bootstrap" Makefile from this project, saves it to the current working
+directory, and then echoes back the name of the saved file. Make reads the bootstrap
```
Comment (kbalk) [inline README.md:145]: I think the sentence "Make read ..." is confusing or unnecessary, especially in the next paragraph when you describe what the bootstrap Makefile is doing.

File: README.md [hunk 5]
```diff
@@ -45,58 +45,140 @@ This project can be utilized one of two ways, via docker or via a Makefile inclu
...
+Makefile and processes its directives and targets.
+
+The bootstrap Makefile is very barebones. Its primary purpose is to bootstrap this
+`tardigrade-ci` project so all the make targets are available. It does this through
+its own `include` directive. The bootstrap include directive processes a shell
```
Comment (kbalk) [inline README.md:150]: The second include could use quotes around it ... if you found this sentence difficult to understand, then you know why.

File: README.md [hunk 6]
```diff
@@ -45,58 +45,140 @@ This project can be utilized one of two ways, via docker or via a Makefile inclu
...
+In **user mode**, the bootstrap Makefile automatically clones the `tardigrade-ci`
+repo to a subdirectory of the calling project, e.g. `/{project}/tardigrade-ci/`.
+This feature is called **auto-init**. In this mode, the `tardigrade-ci` subdirectory
+is considered "owned" by the bootstrap Makefile. The bootstrap Makefile will
+manage and update the clone (through the `include` logic) to ensure the checkout
+matches the git branch or tag specified by the env `TARDIGRADE_CI_BRANCH`. By
+default, this env will resolve to either the version specified in the `Dockerfile`
+(if present) or the `master` branch (if the Dockerfile is missing or is not
+specifying a version). The user may override the default behavior by providing
+the env explicitly, e.g. `TARDIGRADE_CI_BRANCH=foo make help`.
```
Comment (kbalk) [inline README.md:166]: The example value for TARDIGRADE_CI_BRANCH is confusing.  The value is to be a branch, not a command, or am I missing the point?

File: README.md (hunk 4 above)
Comment (lorengordon) [reply README.md:145]: I guess I was trying to explain what the `include` directive actually does. How about this wording?

```
The `include` directive tells `make` to read the specified Makefile and process its directives and targets.
```

Or I can just remove the sentence, if that isn't necessary.

File: README.md (hunk 3 above)
Comment (lorengordon) [reply README.md:155]: What if I just change `whether to clone this project` to `the operating mode (described below)`?

```suggestion
its own `include` directive. The bootstrap `include` directive processes a shell
command with logic to determine the operating mode (described below), and then echoes
back the path to the tardigrade-ci Makefile.

The bootstrap Makefile supports two "operating modes" that we're calling **user mode**
and **developer mode**.

```

File: README.md (hunk 6 above)
Comment (lorengordon) [reply README.md:166]: `foo` was meant to indicate the branch value. How about `{branch}` instead?

```suggestion
the env explicitly, e.g. `TARDIGRADE_CI_BRANCH={branch} make help`.
```

**Comment by lorengordon** [comment]
Thanks @kbalk and @eemperor. Just pushed a commit with the suggestions I put in the comments. Also, if it helps, uou can view the rendered version in my fork: https://github.com/lorengordon/tardigrade-ci/tree/init-updates

File: README.md (hunk 3 above)
Comment (eemperor) [reply README.md]: That works.

File: README.md (hunk 6 above)
Comment (kbalk) [reply README.md]: I'm still confused ... is TARDIGRADE_CI_BRANCH a branch name?  If so, why are you including a command with it?

**Comment by kbalk** [comment]
> Yes, CHANGELOG.md is still needed. We need to be better about updating it though.

And when does that happen?  If nothing else, create a ticket so it's addressed in the future.  It's not the only repo with an outdated CHANGELOG.md.  There are tools available for ensuring sane git commit messages and once you have sane commit messages, it's possible to automate the production of change logs and not have to worry about it again.

File: bootstrap/Makefile.bootstrap [hunk 7]
```diff
@@ -2,19 +2,69 @@ export SHELL = /bin/bash
...
+
+# Macro to clone TARDIGRADE_CI_PROJECT
+# editorconfig-checker-disable
+define $(TARDIGRADE_CI_PROJECT)/clone
+if [[ -d "$(TARDIGRADE_CI_PATH_LOCAL)" ]]; then \
+  echo "Removing existing $(TARDIGRADE_CI_PATH_LOCAL)" >&2 &&\
+  rm -rf "$(TARDIGRADE_CI_PATH_LOCAL)" ;\
+fi && \
+echo "Cloning $(TARGIDRADE_CI_CLONE_URL)?ref=$(TARDIGRADE_CI_BRANCH)..." >&2 &&\
+git clone -c advice.detachedHead=false --depth=1 -b "$(TARDIGRADE_CI_BRANCH)" "$(TARGIDRADE_CI_CLONE_URL)"
```
Comment (kbalk) [inline bootstrap/Makefile.bootstrap:23]: TARGIDRADE_CI_CLONE_URL is consistently misspelled, so it works, but it might be worth correcting because my assumption was that this was an error until I saw it elsewhere.

**Comment by kbalk** [comment]
I am having difficulty testing this.  I started by cloning your pull request into my own ~/plus3it directory and then going to a non-tardigrade-ci directory and creating the .tardigrade-ci from the new bootstrap Makefile.  I also copied the new tardigrade-ci directory into a tardigrade-ci directory in my repo.  The make can't find any targets.  How were you testing it?

File: README.md (hunk 6 above)
Comment (lorengordon) [reply README.md]: `TARDIGRADE_CI_BRANCH` is the env that is used by the bootstrap makefile to set the branch (or tag) for the `init` functionality.

`TARDIGRADE_CI_BRANCH={branch} make help` is shell syntax for setting an env for a single command, without needing to export it into the shell.

File: bootstrap/Makefile.bootstrap (hunk 7 above)
Comment (lorengordon) [reply bootstrap/Makefile.bootstrap:23]: haha good catch, a typo, plus copy pasta of course.

**Comment by lorengordon** [comment]
> I am having difficulty testing this. I started by cloning your pull request into my own ~/plus3it directory and then going to a non-tardigrade-ci directory and creating the .tardigrade-ci from the new bootstrap Makefile. I also copied the new tardigrade-ci directory into a tardigrade-ci directory in my repo. The make can't find any targets. How were you testing it?

Oops, it occurred to me that I needed to write up some test instructions then forgot. I just put it in the PR description.

**Comment by lorengordon** [comment]
>> Yes, CHANGELOG.md is still needed. We need to be better about updating it though.

> And when does that happen?

It's supposed to happen with each release.

**Comment by kbalk** [comment]
> > > Yes, CHANGELOG.md is still needed. We need to be better about updating it though.
> 
> > And when does that happen?
> 
> It's supposed to happen with each release.

But not this one?

File: README.md (hunk 6 above)
Comment (kbalk) [reply README.md]: That's obvious now, but I couldn't see it before.  I'll look at the text again on Monday as this was confusing.

**Comment by kbalk** [comment]
> > I am having difficulty testing this. I started by cloning your pull request into my own ~/plus3it directory and then going to a non-tardigrade-ci directory and creating the .tardigrade-ci from the new bootstrap Makefile. I also copied the new tardigrade-ci directory into a tardigrade-ci directory in my repo. The make can't find any targets. How were you testing it?
> 
> Oops, it occurred to me that I needed to write up some test instructions then forgot. I just put it in the PR description.

Actually, your change to Makefile to move 'shell' before 'test' solved my problem.  But I will also try to test it the ways you've indicated

File: README.md (hunk 6 above)
Comment (kbalk) [reply README.md]: The earlier part of the paragraph references the envvar TARDIGRADE_CI_BRANCH, so I don't know why I would think that you'd need to just show how to set that envvar.  It is dense reading and you can get lost in the weeds, so maybe change:

```
The user may override the default behavior by providing the env explicitly, e.g., ...
```

with

```
The user may override the default behavior by providing the env explicitly before invoking make, e.g., ...
``` 

It's a bit lame, but I kept missing the obvious.

Comment (lorengordon) [reply README.md]: Roger. Perhaps I'm trying too hard. How about we just remove that sentence? If we need to, we can add a separate section on the bootstrap envs.

Comment (kbalk) [reply README.md]: I would keep the sentence and add the "before invoking make".  

You have a lot of envvars.  Most people won't need to know that level of detail, i.e., all the envvars.  @eemperor - what do you think?

**Comment by kbalk** [comment]
> Testing user mode:
> 
 ...
>    $ TARDIGRADE_CI_ORG=lorengordon TARDIGRADE_CI_BRANCH=init-updates make help

The `make init` command will pull in the tardigrade-ci master branch, so there's probably an envvar missing in this description.  But since this info isn't officially documented, it's not an issue.

**Comment by lorengordon** [comment]
> The `make init` command will pull in the tardigrade-ci master branch, so there's probably an envvar missing in this description. But since this info isn't officially documented, it's not an issue.

To use a fork/branch, it is necessary to _consistently_ use the envs `TARDIGRADE_CI_ORG` and `TARDIGRADE_CI_BRANCH`. Every invocation of `make`. Or yes, it will revert to the default value and pull down the plus3it+master version.

```
$ TARDIGRADE_CI_ORG=lorengordon TARDIGRADE_CI_BRANCH=init-updates make init
Removing existing /c/Users/x/Git/test/tardigrade-ci
Cloning https://github.com/lorengordon/tardigrade-ci.git?ref=init-updates...  # <--- `init` is using the specified fork/branch
Cloning into 'tardigrade-ci'...
remote: Enumerating objects: 63, done.
remote: Counting objects: 100% (63/63), done.
remote: Compressing objects: 100% (50/50), done.
remote: Total 63 (delta 24), reused 27 (delta 10), pack-reused 0
Receiving objects: 100% (63/63), 56.23 KiB | 928.00 KiB/s, done.
Resolving deltas: 100% (24/24), done.
git -C "/c/Users/x/Git/test/tardigrade-ci" log -1 --oneline
f8e9292 (grafted, HEAD -> init-updates, origin/init-updates) Updates changelog for all missing versions  # <--- confirms the specified fork/branch
[init]: Completed successfully!
```

**Comment by kbalk** [comment]
I've been testing in a clone of the security-tool repo and followed step 3 of your "Testing user mode" instructions.  I've been trying various combinations and this combo failed:

```
TARDIGRADE_CI_ORG=lorengordon TARDIGRADE_CI_BRANCH=init-updates make clean
TARDIGRADE_CI_ORG=lorengordon TARDIGRADE_CI_BRANCH=init-updates make docker/run target=python/format
Cloning https://github.com/lorengordon/tardigrade-ci.git?ref=init-updates...
Cloning into 'tardigrade-ci'...
remote: Enumerating objects: 63, done.
remote: Counting objects: 100% (63/63), done.
remote: Compressing objects: 100% (50/50), done.
remote: Total 63 (delta 24), reused 27 (delta 10), pack-reused 0
Receiving objects: 100% (63/63), 56.23 KiB | 3.31 MiB/s, done.
Resolving deltas: 100% (24/24), done.
.tardigrade-ci:45: /home/kbalk/plus3it/tardigrade-ci/Makefile: No such file or directory
make: *** No rule to make target `/home/kbalk/plus3it/tardigrade-ci/Makefile'.  Stop.
```

The `make clean` wiped out the underlying .tardigrade-ci and the `make docker/run` didn't know to recreate it.  It's a non-normal scenario, but ...

**Comment by lorengordon** [comment]
> I've been trying various combinations and this combo failed:

The `docker/run` target requires a special test setup, particularly when the targets and the Dockerfile have changed significantly. You have to pre-create the updated docker image, noting the image ID, and set the env `IMAGE_ID={image_id}` when invoking `docker/run`. And even that might not be working entirely right now, in user mode. I _think_ we'd have to expose the envs `TARDIGRADE_CI_...` to the docker container, or the auto-init logic will resolve to the plus3it+master branch default...

**Comment by kbalk** [comment]
> > I've been trying various combinations and this combo failed:
> 
> The `docker/run` target requires a special test setup, particularly when the targets and the Dockerfile have changed significantly. You have to pre-create the updated docker image, noting the image ID, and set the env `IMAGE_ID={image_id}` when invoking `docker/run`. And even that might not be working entirely right now, in user mode. I _think_ we'd have to expose the envs `TARDIGRADE_CI_...` to the docker container, or the auto-init logic will resolve to the plus3it+master branch default...

If you're saying this would work if we weren't testing a new variant of targdigrade-ci, then I would ignore this case.  The Makefile and Makefile.bootstrap is already too complicated.

Wait a sec, my /home/kbalk/plus3it/tardigrade-ci/ directory was wiped out ...

**Comment by kbalk** [comment]
> > > I've been trying various combinations and this combo failed:
> > 
> > 
> > The `docker/run` target requires a special test setup, particularly when the targets and the Dockerfile have changed significantly. You have to pre-create the updated docker image, noting the image ID, and set the env `IMAGE_ID={image_id}` when invoking `docker/run`. And even that might not be working entirely right now, in user mode. I _think_ we'd have to expose the envs `TARDIGRADE_CI_...` to the docker container, or the auto-init logic will resolve to the plus3it+master branch default...
> 
> If you're saying this would work if we weren't testing a new variant of targdigrade-ci, then I would ignore this case. The Makefile and Makefile.bootstrap is already too complicated.
> 
> Wait a sec, my /home/kbalk/plus3it/tardigrade-ci/ directory was wiped out ...

Yes, it's repeatable -- I can wipe out my tardigrade-ci directory.  *sigh*   

I'm done bumbling around ... outside of needing to be very careful testing changes to tardigrade-ci, I didn't find other errors.  I would like to see the minor change to the README.md, but otherwise it looks good.

**Comment by lorengordon** [comment]
> Yes, it's repeatable -- I can wipe out my tardigrade-ci directory. sigh

Please provide exact reproduction steps, preferably a complete command history and output logs. Previously this was possible, but part of the changes here were to improve the safeties of the `clean` target so this could no longer happen. For one, the `clean` target is now using `TARDIGRADE_CI_PATH_LOCAL`, which should only resolve to a subdirectory of the calling project.

**Comment by kbalk** [comment]
Three commands are needed:

```
TARDIGRADE_CI_ORG=lorengordon TARDIGRADE_CI_BRANCH=init-updates make clean
TARDIGRADE_CI_ORG=lorengordon TARDIGRADE_CI_BRANCH=init-updates make docker/clean
TARDIGRADE_CI_ORG=lorengordon TARDIGRADE_CI_BRANCH=init-updates make docker/run target=python/format
 ... snipped ...
[docker/run]: Running docker image
docker run --rm \
-v "/home/kbalk/plus3it/security-tools/:/workdir/" \
-v "/home/kbalk/plus3it/tardigrade-ci/:/tardigrade-ci/" \.      <---- WHY?
-v "/home/kbalk/.aws/:/root/.aws/" \
-e AWS_DEFAULT_REGION=us-east-1 \
-e AWS_PROFILE=dicelab-dev \
-w /workdir/ \
--entrypoint make \
security-tools:latest python/format
Removing existing /tardigrade-ci
rm: cannot remove '/tardigrade-ci': Device or resource busy
.tardigrade-ci:45: /tardigrade-ci/Makefile: No such file or directory
make: *** No rule to make target '/tardigrade-ci/Makefile'.  Stop.
make: *** [docker/run] Error 2
```

Afterwards, I'll have a ~/plus3it/tardigrade-ci directory, but it will be empty.

**Comment by lorengordon** [comment]
> -v "/home/kbalk/plus3it/tardigrade-ci/:/tardigrade-ci/" \.      <---- WHY?

That supports developer mode, so tardigrade-ci work you have in-flight locally is available to the container.

I think I see what's happening. The auto-init inside the docker container is conflicting with the logic for developer mode. I'll fix it.

File: README.md (hunk 6 above)
Comment (lorengordon) [reply README.md]: @kbalk I've updated the wording.

**Comment by lorengordon** [comment]
> I think I see what's happening. The auto-init inside the docker container is conflicting with the logic for developer mode. I'll fix it.

@kbalk I pushed a couple commits that I think fix the issue where it could wipe out the contents of an external `tardigrade-ci` repo. There was a bug in the use case where auto-init, "developer mode", and docker/run are all involved. Now it explicitly checks if it is running in docker and skips the auto-init.

File: README.md (hunk 6 above)
Comment (kbalk) [reply README.md]: Thank you.

**Review by kbalk — APPROVED** [review]
I didn't find any new errors.  I didn't perform a very systematic test, but I'm hoping what I did was sufficient.  So it looks good for a general release.
